src/server/ldap/client.ts

```diff
@@ -0,0 +1,128 @@
+import * as ldap from 'ldapjs';
+import * as express from 'express';
+import {EventEmitter} from 'events';
+import {User} from '../../lib/model';
+
+interface ILdapConfig {
+    server: {
+        url: string,
+    };
+    base: string;
+    search: ldap.Client.SearchOptions;
+}
+let ldapConfig: ILdapConfig = {
+    server: {
+        url: 'ldap://0.0.0.0:1389',
+    },
+    base: 'o=eduxored',
+    search: {
+        filter: '(login=1)',
+        scope: 'base',
+    },
+};
+
+let clientOpts: ldap.Client.ClientOptions = {
+    url: 'ldap://0.0.0.0:1389',
+};
+
+/**
+ * Searches for a user by login, if it was found,
+ * compares its password with the one in the request.
+ */
+export default async function resolveUser(login: string, password: string,
+                                    cb: (error: string, user: User) => void) {
+    let error: string;
+    let user: User;
+    let client: ldap.Client;
+    if (!login || !password) {
+        cb(ldap.LDAP_INAPPROPRIATE_AUTHENTICATION + ':null:No username or password:null', null);
+        return;
+    };
+    client = await makeClient((err: string) => {
+        if (err) {
+            error = err;
+        }
+    });
+    if (error) {
+        return cb(error, null);
+    }
+    let searchOpts: ldap.Client.SearchOptions = ldapConfig.search;
+    searchOpts.filter = '(login=' + login + ')';
+
+    try {
+    await client.search('cn=' + login + ', ' + ldapConfig.base, searchOpts,
+                                            (err: ldap.LDAPError, res: EventEmitter) => {
+        res.on('searchEntry', async (entry: any) => {
+            try {
+            await compare(client, entry.dn, 'pwdhash', password, async (err: string) => {
+                if (err) {
+                    error = err;
+                } else {
+                    try {
+                    user = await makeUserFromEntry(entry);
+                    } catch (error) {
+                        return cb(error, null);
+                    }
+                }
+                return cb(error, user);
+            });
+            } catch (error) {
+                return cb(error, null);
+            }
+        });
+        res.on('error', (err: ldap.LDAPError) => {
+            return cb(err.code + ':' + err.dn + ':' + err.message + ':' + err.name, null);
+        });
+    });
+    } catch (error) {
+        cb(error, null);
+    }
+}
+
+/**
+ * Makes common client.
+ */
+async function makeClient(cb: (error: string) => void): Promise<ldap.Client> {
```

Comment:
also since you return a promise, you could remove error `callback` and write the following code instead:

```js
return new Promise<ldap.Client>((resolve, reject) => {
  let client: ldap.Client = ldap.createClient(clientOpts);
    client.on('error', (err: ldap.LDAPError) => {
        reject(err.code + ':' + err.dn + ':' + err.message + ':' + err.name);
     });
     resolve(client);
});
```